### ISSUE TITLE
ReactiveMongoClient performance on large result sets

### ISSUE THREAD

**Issue by tomas1885** [quarkusio/quarkus#36177]
### Describe the bug

When using ReactiveMongoClient for getting large result sets there's a huge performance penalty (orders of magnitude) compared to using the blocking client. 

The issue stems from the usage of `io.quarkus.mongodb.impl.Wrappers#toMulti, since it executes each `onItem` on the current Vertx context if possible.

The difference for our use case is **2s** vs **28s**. 
I understand the need to emit each item on the calling Vertx context, we need to find a different way as the overhead is just too high. 
Simple using `AdaptersToFlow.publisher` without the call to vertx results in low latency but the item is not emitted on the calling Vertx context. 
The only way I found to be able to keep the item emitting on the Vertx Context without suffering from the performance penalty is to use `AdaptersToFlow.publisher` and wrap it in a replying Multi with emitOn either a `MutinyHelper.executor(currentContext)` or simply `emitOn(cmd->ctx.runOnContext(cmd))`. 
I'm not sure if there are any side effects and there might be better ways of solving this issue, but for now we're stuck on using the underlaying MongoClient (unwrap) with this custom solution. 

Any help would be appreciated

### Expected behavior

ReactiveMongoClient performance should be up par with the blocking client or with using the Reactive Mongo driver from mongodb.

### Actual behavior

The performance for large result set is worse by orders of magnitude.

### How to Reproduce?

Query a large collection with large result set.

### Output of `uname -a` or `ver`

Darwin Kernel Version 22.6.0: Fri Sep 15 13:41:28 PDT 2023; root:xnu-8796.141.3.700.8~1/RELEASE_ARM64_T6020 arm64

### Output of `java -version`

openjdk version "17.0.7" 2023-04-18

### Quarkus version or git rev

3.3.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Build time:   2023-08-17 07:06:47 UTC Revision:     8afbf24b469158b714b36e84c6f4d4976c86fcd5  Kotlin:       1.9.0 Groovy:       3.0.17 Ant:          Apache Ant(TM) version 1.10.13 compiled on January 4 2023 JVM:          17.0.7 (Eclipse Adoptium 17.0.7+7) OS:           Mac OS X 13.6 aarch64

**Comment by tomas1885** [comment]
I tried using the AsyncProfiler but I don't think this is the profile you're looking for, do you have any specific modes (profiler flags) you want me to use?
Because providing a sample app would require filling a local MongoDB with lot's of data

**Comment by geoand** [comment]
If the difference in the two styles is so large,I would assume that a simple CPU profile would show it (and I am pretty sure a memory profile would also show huge allocations of Mutiny types).

That said, @franz1981 is the expert at async-profiler settings :)

**Comment by tomas1885** [comment]
[profiling.zip](https://github.com/quarkusio/quarkus/files/12737846/profiling.zip)
Here's a Zip containing the profile html (I followed the instructions) of the reactive and blocking calls accordingly, If you want it in a different format please let me know

**Comment by franz1981** [comment]
Got a couple of questions:
- what's the duration of the profiling data?
- have you added the  -XX:+UnlockDiagnosticVMOptions -XX:+DebugNonSafepoints JVM args? I cannot see many expected inlined methods, so I suppose it was missing.
- how many cores the server is using?

**Comment by tomas1885** [comment]
I did add these JVM args, I also didn't see much valuable information there, so I took the time to write a simple reproducer. Attached here as a zip file, there are 3 endpoint, one for blocking, one for rea
[mongoreproducer.zip](https://github.com/quarkusio/quarkus/files/12738188/mongoreproducer.zip)
ctive and another one with the hacky workaround. You can clearly see the issue.

**Comment by geoand** [comment]
Thanks a lot, I'll have when I can.

**Comment by cescoffier** [comment]
(Adding @jponge)

Thanks for this report. The mongo facade was implemented a very long time ago (before we had mutiny, actually). Many improvements can be done, and I think you found one. Using `emitOn` and using a trampoline would provide much better performances. 

WDYT @jponge ?

**Comment by jponge** [comment]
Perhaps []ContextAwareScheduler(https://smallrye.io/smallrye-mutiny-vertx-bindings/3.6.0/apidocs/io/smallrye/mutiny/vertx/core/ContextAwareScheduler.html) could be used to simplify pushing back execution on the correct Vert.x duplicated context

**Comment by jponge** [comment]
Ah my bad, this doesn't require a scheduler 🤦

**Comment by tomas1885** [comment]
FYI, for now, as a workaround replace aggregate and find calls with something like this:
`		
		var collection =db.unwrap.getCollection(PayoutsMongoRepository.COLLECTION);
Context context = Vertx.currentContext();
		if (context != null) {
			return Multi.createBy().replaying().ofMulti(Multi.createFrom().publisher(AdaptersToFlow.publisher(collection.aggregate(pipeline,clazz))))
					.emitOn(cmd->context.runOnContext(x->cmd.run()));
		} else {
			return Multi.createFrom().publisher(AdaptersToFlow.publisher(collection.aggregate(pipeline,clazz)));
		}`

This doesn't seem to impact the performance and the item emission still happens on the same calling Vertx Context

**Comment by gsmet** [comment]
@jponge I'm having a look at old issues and it looks like this one could be addressed relatively easily?

**Comment by jponge** [comment]
I'm not sure why the performant solution is to replay item, because this requires allocations and it's really about shuffling data around in memory.

**Comment by gsmet** [comment]
@jponge not saying the workaround is what we should do but I thought you might have the expertise to make things better in a relatively short time, from what I read from the discussions.

I tried the reproducer and I got:
```
2024-11-21 16:07:44,150 INFO  [com.exa.ExampleResource] (executor-thread-1) blocking took 235 ms
2024-11-21 16:07:46,269 INFO  [com.exa.ExampleResource] (vert.x-eventloop-thread-1) reactive took 1373 ms
```

So that's a nearly 6x slowdown on my laptop.
That's with the reproducer updated to Quarkus 3.16.4.
And this is with a rather suboptimal blocking pattern as the list of documents is copied.

I'm perfectly fine with reactive not being as fast as the blocking counterpart in this use case but the 6x slow down seems like something we should have a look at.

I attached the updated reproducer:

[mongodb-reproducer-updated.zip](https://github.com/user-attachments/files/17847964/mongodb-reproducer-updated.zip)

Running `./gradlew quarkusDev` and then loading http://localhost:8080/dummy/blocking and http://localhost:8080/dummy/reactive will get you the results.

**Comment by jponge** [comment]
Thanks @gsmet I will have a look

**Comment by cescoffier** [comment]
The solution is relatively simple. We should batch emission instead of writing each item on the context individually.

**Comment by jponge** [comment]
The reactive client calls get all routed through `io.quarkus.mongodb.impl.Wrappers`, which adapts to a `Multi` but most importantly uses `emitOn` to dispatch to a Vert.x context and have context propagation, so I'm not exactly sure there's an easy fix.

**Comment by cescoffier** [comment]
I said 'relatively' :-) 

Now it's not trivial. Basically, we would need to fork the emitOn operator. If I'm not mistaken, there is an internal queue. Instead of dispatching item individually, we could clear the queue in one pass. Maybe we could add a threshold to avoid monopolization.

**Comment by jponge** [comment]
If you want efficiency and loose context propagation, you need to do something like:

```java
FindPublisher<DummyDoc> findPublisher = clientReactive.unwrap().getDatabase(DummyData.DB_NAME)				
  .getCollection(DummyData.COLLECTION_NAME, DummyDoc.class)
  .find();

return Multi.createFrom().publisher(AdaptersToFlow.publisher(findPublisher))
  .collect().asList().onItem().invoke(res -> {
    logger.info("reactive took " + (System.currentTimeMillis() - start) + " ms");
});
```

The items get emitted from the Mongo driver managed threads, which we don't have any control on.

The slow version with context propagation is like this:

```java
Multi<DummyDoc> find = clientReactive.getDatabase(DummyData.DB_NAME)
  .getCollection(DummyData.COLLECTION_NAME, DummyDoc.class)
  .find();

return find.collect().asList()
  .onItem().invoke(res -> {
    logger.info("reactive took " + (System.currentTimeMillis() - start) + " ms");
});
```

or even like this if you want to `.unwrap()` to reduce the number of thread hops:

```java
FindPublisher<DummyDoc> findPublisher = clientReactive.unwrap().getDatabase(DummyData.DB_NAME)
  .getCollection(DummyData.COLLECTION_NAME, DummyDoc.class)
  .find();

return Multi.createFrom().publisher(AdaptersToFlow.publisher(findPublisher))
  .emitOn(cmd -> context.runOnContext(x -> cmd.run()))
  .collect().asList().onItem().invoke(res -> {
     logger.info("reactive took " + (System.currentTimeMillis() - start) + " ms");
});
```

I have a suspicion that the "workaround" with replays of `Multi` is not measuring exactly what is assumed to be measured, but more something equivalent to this, which has most of the work on the Mongo threads, then hops back to a Vert.x context:

```java
FindPublisher<DummyDoc> findPublisher = clientReactive.unwrap().getDatabase(DummyData.DB_NAME)
  .getCollection(DummyData.COLLECTION_NAME, DummyDoc.class)
  .find();

return Multi.createFrom().publisher(AdaptersToFlow.publisher(findPublisher))				
  .collect().asList()
  .emitOn(cmd -> context.runOnContext(x -> cmd.run()))
  .onItem().invoke(res -> {
     logger.info("reactive took " + (System.currentTimeMillis() - start) + " ms");
});
```

To me this boils down to the Mongo client having its own network stack which we have no control on, forcing us to design a reactive facade that does too many thread hops in such a use-case. And again, this is due to the assumption that we want auto-magic context propagation from a different threading model, if not then we get no extra latency penalty.

@cescoffier perhaps we can have a pluggable emitter inside `Wrappers` that does some batching, but it sounds easier said than done 😉  (w.r.t. backpressure, etc)

**Comment by jponge** [comment]
Just for fun, I might try something like this and see if it works 🤣 

```java
 return Multi.createFrom().publisher(AdaptersToFlow.publisher(publisher))
                    .group().intoLists().of(10)
                    .emitOn(command -> context.runOnContext(x -> command.run()))
                    .onItem().transformToMultiAndConcatenate(items -> Multi.createFrom().iterable(items));
```

**Comment by jponge** [comment]
Batching does have an impact on some artificial test that I created (I could not easily test with this project being Gradle-based, and the POM I created had issues to enter `quarkus:dev`)

**Comment by jponge** [comment]
What we could do is having another `Multi` wrapper that does batching (grouping then flattening is ok) and offer it for some operations. 

We could have some Quarkus-level configuration of the driver for:
- enabling batching or not and keep the old behavior
- when batching is enabled, we'd use `.group().intoLists().of(int size, Duration maximumDelay)`
  - the batch sizes
  - the flush delay, because otherwise some data could be stuck due to demand

That being said we always get back to the problem that when a reactive driver comes with a custom networking stack there are such limitations because of the need to hop on a Vert.x context, or else use `.unwrap()` which is good to have but you need to understand what you are doing.

**Comment by gsmet** [comment]
I dunno, frankly if the switching to a `Multi` is border line unusable in this case, I would be in favor of dropping this code entirely.

It seems wrong to suggest that people this is a supported feature, while you will have issues any time your record set is huge.
Now what I don't know is which things will get more complex/impossible if we drop this.

Now, where I'm worried is that it seems we could have to pay this price in some other extensions? If so, I'm a bit worried about this.

**Comment by cescoffier** [comment]
The Mongo driver is unique here. We had to implement our own facade to avoid other, bigger issues related to the internals of the driver. Now, yes, that’s the cost to pay. 

We can improve but it’s not a small effort.

**Comment by jponge** [comment]
> Now, where I'm worried is that it seems we could have to pay this price in some other extensions? If so, I'm a bit worried about this.

No need to worry, this is due the the peculiarities of this driver and how it "streams".

**Comment by jponge** [comment]
Quick update: the Mutiny `emitOn` is doing the right thing, aka it does ensure thread piggy-backing rather than dispatching to an executor for each item.

**Comment by jponge** [comment]
Another quick update: I've run some profling of the `emitOn` operator on artificial CPU-bound workloads, in itself it doesn't have an impact of the order of magnitude that's being observed in the reproducer, even when doing multiple `emitOn` steps.

**Comment by jponge** [comment]
Fun fact: there is actually no performance issue in Quarkus.

More on that tomorrow, I need to catch up on 😴

**Comment by jponge** [comment]
### TL;DR

- Short answer: there is no performance issue in Quarkus, Vert.x or Mutiny.
- Not so short answer: the Mongo driver doesn't do proper batching with the default parameters, and passing a batch size parameters does fix the observed latency issues. I will look into the Quarkus documentation to possibly update the example.

### Reproducer

I wrote another and cleaner reproducer at https://github.com/jponge/reproducer-quarkus-mongo-reactive-perf-batchsize 

The reproducer initializes a dataset for 50k entries and just once, unlike the earlier reproducer (in dev mode the latency was getting worse and worse because reloading could easily pile more and more data when the database was already up and running).

3 endpoints are exposed.

#### Imperative / blocking endpoint (`/api/all`)

The REST endpoint is:

```java
@GET
@Path("all")
@Produces(MediaType.APPLICATION_JSON)
public List<Message> syncFetch() {
    return messageService.fetchAllSync();
}
```

The data fetching code is:

```java
public List<Message> fetchAllSync() {
    long start = System.currentTimeMillis();
    logger.info("Fetching all messages synchronously");
    ArrayList<Message> messages = mongoClient
            .getDatabase(DB_NAME)
            .getCollection(COLLECTION_NAME, Message.class)
            .find()
            .into(new ArrayList<>());
    logger.info(messages.size() + " messages fetched in " + (System.currentTimeMillis() - start) + "ms");
    return messages;
}
```

One-shot request trace:

```
2024-12-04 09:57:39,329 INFO  [dat.MessageService] (executor-thread-1) Fetching all messages synchronously
2024-12-04 09:57:39,457 INFO  [dat.MessageService] (executor-thread-1) 50000 messages fetched in 128ms
```

#### Reactive / non-blocking endpoint without batching (`/api/async/all-bad`)

The REST endpoint is:

```java
@GET
@Path("async/all-bad")
@Produces(MediaType.APPLICATION_JSON)
public Uni<List<Message>> asyncFetchButBadly() {
    return messageService.fetchAllAsyncButBadly();
}
```

The data fetching code is:

```java
public Uni<List<Message>> fetchAllAsyncButBadly() {
    long start = System.currentTimeMillis();
    logger.info("Fetching all messages asynchronously without batching");
    return reactiveMongoClient
            .getDatabase(DB_NAME)
            .getCollection(COLLECTION_NAME, Message.class)
            .find()
            .collect().asList()
            .onItem().invoke(messages -> logger.info(messages.size() + " messages fetched in " + (System.currentTimeMillis() - start) + "ms"));
}
```

One-shot request trace (about `10x` the imperative case):

```
2024-12-04 10:10:26,080 INFO  [dat.MessageService] (vert.x-eventloop-thread-1) Fetching all messages asynchronously without batching
2024-12-04 10:10:27,119 INFO  [dat.MessageService] (vert.x-eventloop-thread-1) 50000 messages fetched in 1039ms
```

#### Reactive / non-blocking endpoint with batching (`/api/async/all`)

The REST endpoint is:

```java
@GET
@Path("async/all")
@Produces(MediaType.APPLICATION_JSON)
public Uni<List<Message>> asyncFetch() {
    return messageService.fetchAllAsync();
}
```

The data fetching code is:

```java
public Uni<List<Message>> fetchAllAsync() {
    long start = System.currentTimeMillis();
    logger.info("Fetching all messages asynchronously with batching");
    return reactiveMongoClient
            .getDatabase(DB_NAME)
            .getCollection(COLLECTION_NAME, Message.class)
            .find(new FindOptions().batchSize(4096))
            .collect().asList()
            .onItem().invoke(messages -> logger.info(messages.size() + " messages fetched in " + (System.currentTimeMillis() - start) + "ms"));
}
```

One-shot request trace (marginally slower to the imperative case in the range of `+10ms`):

```
2024-12-04 10:12:39,006 INFO  [dat.MessageService] (vert.x-eventloop-thread-1) Fetching all messages asynchronously with batching
2024-12-04 10:12:39,145 INFO  [dat.MessageService] (vert.x-eventloop-thread-1) 50000 messages fetched in 139ms
```

### Load testing

I used `hey`, a HTTP load testing tool that does not suffer from _coordinated omission_.

We observe a more consistent latency control of the reactive code (with the driver batching hint) compared to the imperative counterpart.
This is consistent with what is to be expected from using a reactive API.

The reactive code with driver defaults for batching (`0`) results confirms a performance penalty, but unlike on a one-shot request where we have a `10x` latency factor, the service under load provides a `4x` latency hit.

#### Imperative / blocking endpoint (`/api/all`)

```
Summary:
  Total:	5.6766 secs
  Slowest:	2.9303 secs
  Fastest:	0.1776 secs
  Average:	1.3190 secs
  Requests/sec:	35.2326
  

Response time histogram:
  0.178 [1]	|■
  0.453 [12]	|■■■■■■■■■■■
  0.728 [34]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  1.003 [42]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  1.279 [23]	|■■■■■■■■■■■■■■■■■■■■■■
  1.554 [16]	|■■■■■■■■■■■■■■■
  1.829 [24]	|■■■■■■■■■■■■■■■■■■■■■■■
  2.104 [10]	|■■■■■■■■■■
  2.380 [17]	|■■■■■■■■■■■■■■■■
  2.655 [14]	|■■■■■■■■■■■■■
  2.930 [7]	|■■■■■■■


Latency distribution:
  10% in 0.5436 secs
  25% in 0.7559 secs
  50% in 1.1801 secs
  75% in 1.7806 secs
  90% in 2.4021 secs
  95% in 2.6262 secs
  99% in 2.8437 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0022 secs, 0.1776 secs, 2.9303 secs
  DNS-lookup:	0.0014 secs, 0.0000 secs, 0.0063 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0026 secs
  resp wait:	1.2404 secs, 0.1577 secs, 2.9200 secs
  resp read:	0.0763 secs, 0.0009 secs, 0.4941 secs

Status code distribution:
  [200]	200 responses
```

#### Reactive / non-blocking endpoint with batching (`/api/async/all`)

```
Summary:
  Total:	6.1900 secs
  Slowest:	2.4456 secs
  Fastest:	1.0361 secs
  Average:	1.4472 secs
  Requests/sec:	32.3101
  

Response time histogram:
  1.036 [1]	|■
  1.177 [7]	|■■■■
  1.318 [45]	|■■■■■■■■■■■■■■■■■■■■■■■■■■
  1.459 [68]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  1.600 [42]	|■■■■■■■■■■■■■■■■■■■■■■■■■
  1.741 [20]	|■■■■■■■■■■■■
  1.882 [5]	|■■■
  2.023 [11]	|■■■■■■
  2.164 [0]	|
  2.305 [0]	|
  2.446 [1]	|■


Latency distribution:
  10% in 1.2198 secs
  25% in 1.3105 secs
  50% in 1.4129 secs
  75% in 1.5460 secs
  90% in 1.7179 secs
  95% in 1.8861 secs
  99% in 2.0135 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0009 secs, 1.0361 secs, 2.4456 secs
  DNS-lookup:	0.0004 secs, 0.0000 secs, 0.0019 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0006 secs
  resp wait:	1.4045 secs, 1.0097 secs, 2.4221 secs
  resp read:	0.0418 secs, 0.0155 secs, 0.3005 secs

Status code distribution:
  [200]	200 responses
```

#### Reactive / non-blocking endpoint without batching (`/api/async/all-bad`)

```
Summary:
  Total:	32.0163 secs
  Slowest:	8.5533 secs
  Fastest:	7.3522 secs
  Average:	7.9051 secs
  Requests/sec:	6.2468
  

Response time histogram:
  7.352 [1]	|■
  7.472 [3]	|■■
  7.592 [9]	|■■■■■■■
  7.713 [24]	|■■■■■■■■■■■■■■■■■■■■
  7.833 [43]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  7.953 [49]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  8.073 [25]	|■■■■■■■■■■■■■■■■■■■■
  8.193 [23]	|■■■■■■■■■■■■■■■■■■■
  8.313 [12]	|■■■■■■■■■■
  8.433 [8]	|■■■■■■■
  8.553 [3]	|■■


Latency distribution:
  10% in 7.6312 secs
  25% in 7.7736 secs
  50% in 7.8732 secs
  75% in 8.0487 secs
  90% in 8.2563 secs
  95% in 8.3255 secs
  99% in 8.4809 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0012 secs, 7.3522 secs, 8.5533 secs
  DNS-lookup:	0.0006 secs, 0.0000 secs, 0.0025 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0006 secs
  resp wait:	7.8818 secs, 7.3304 secs, 8.5312 secs
  resp read:	0.0221 secs, 0.0199 secs, 0.0556 secs

Status code distribution:
  [200]	200 responses

```

### Why is that?

The Mongo driver has a batching configuration parameter in several places.
While digging into the source code, I figured out that the default value is `0`, meaning _"the driver will do what it thinks is best"_.

It is very clear that in the imperative code the driver does the right thing, but in the reactive case we need to take control, as in:

```java
.find(new FindOptions().batchSize(4096))
```

I have no definite answer on why the driver doesn't use a big enough batch size by default in reactive, which results in more frequent Vert.x context scheduling tasks.
As mentionned earlier the Mutiny `emitOn` operator itself does the right thing: when a fast-paced set of emissions happen then dispatching is made by piggy-backing on the scheduled thread rather than scheduling each item emission.
There is a neglictible performance hit here, even when chaining multiple `emitOn` operators in an artificial micro-benchmark.
The only way to get a performance hit here is when items are emitted slowly, or in small batches, or even one-by-one, offering a chance for I/O to intervene and defeat a piggy-backing opportunity.

An easy explanation is that the driver simply doesn't do a good job and that we can't trust its default configuration.
Another explanation could be in the fact that the driver uses Reactor to expose a Reactive Streams implementation.

As we discussed in https://dl.acm.org/doi/10.1145/3486605.3486788, Reactor uses several non-standard tricks including pre-fetching requests, but they vanish as soon as a reactive pipeline is not end-to-end Reactor, falling back to slower but Reactive Streams compliant operator implementations.
It could be the driver expecting Reactor to do some pre-fetching, or the Reactor segment doing small batches on its own, or even the 2 libraries behaving incorrectly while the subscriber (Quarkus Rest) does a `Long.MAX_VALUE` request here.

### What's next?

I will work on a documentation update to encourage explict batching hints.

**Comment by tomas1885** [comment]
Hi @jponge, @gsmet 

I appreciate you picking this issue up and your through analysis.
After further debugging, it seems that the issue stems from the MultiEmitOnProcessor which has a hardcoded limit of 16, which in turn caps the demand of the publisher to 16 at a time. This is observed in the BatchCursorFlux class, which controls the demand of the stream. This is why when using a replaying Multi or .onOverflow().bufferUnconditionally() this demand is than set to the maximum.
Doing something like this also achieves comparable results:
Multi.createFrom().publisher(AdaptersToFlow.publisher(clientReactive.getDatabase(DummyData.DB_NAME).unwrap().getCollection(DummyData.COLLECTION_NAME, DummyDoc.class)
				.find().batchSize(1024*1024))).onOverflow().bufferUnconditionally().emitOn(cmd-> context.runOnContext(x->cmd.run())).collect().asList();
Setting the Batch size also bypasses the issue as it overrides the demand of the subscriber, but I think the root issue here is Mutiny which caps the stream demand to 16, which is propagated the the Flux implementation of Mongo.

**Comment by jponge** [comment]
Many thanks @tomas1885 for spotting this, I had completely missed that `limit` parameter, we should make it configurable because that can indeed make for a bottleneck.

That being said using a replay operator is not what I would recommend, because you are buffering a large result set and this might quickly cause out of memory issues.

In any case I think an explicit batching setting to the driver is still to be recommended, because the driver implementation stills does in-memory batching from what I understand instead of having any real back-pressure mechanism to propagate to the server.

**Comment by tomas1885** [comment]
Thanks @jponge, I understand that using the replay operator (or any other operator that overrides the limit/requested) is subpar, and batching is still the recommended approach. Just one note, It seems that the driver can actually handle backpressure, it allows setting a backpressure strategy, but more importantly it can batch fetch according to the demand defined by the sink. This open the possibility to intelligently "stream" batches back to the consumer and pace the demand as it uses sink.requestedFromDownstream() to calculate the current cursor iteration batch size (which fetches seems to fetch from the server). Not sure yet how it can be incorporated with Mutiny as it relies on the Reactor semantics. But just to keep in mind.

**Comment by jponge** [comment]
@tomas1885 If the driver relies on the demand (as per the RS protocol, not the proprietary extensions in Reactor) then it should work well with Mutiny. I'm likely going to test alternative implementations of `emitOn` so explicit batching might not be needed in such cases.
